File: C++/Code/GoaOnline.cpp
#include <iostream>
#include <string>

using namespace std;

class User {
private:
    static int userCounter;
    int userId;
    string name;
    string email;
    string phoneNumber;
    string address;
    string userType;

public:
    User(const string &name, const string &email, const string &phoneNumber, const string &address, const string &userType)
        : name(name), email(email), phoneNumber(phoneNumber), address(address), userType(userType) {
        userId = userCounter++;
    }

    int getUserId() const {
        return userId;
    }

    void registerUser() const {
        cout << "User registered successfully: " << name << " (User ID: " << userId << ")" << endl;
    }

    bool login(const string &email, const string &phoneNumber) const {
        if (this->email == email && this->phoneNumber == phoneNumber) {
            cout << "User logged in successfully: " << name << endl;
            return true;
        } else {
            cout << "Login failed. Incorrect email or phone number." << endl;
            return false;
        }
    }

    void updateProfile(const string &newAddress) {
        address = newAddress;
        cout << "Profile updated with new address: " << address << endl;
    }

    void viewServices() const {
        cout << "Viewing available services for user: " << name << endl;
    }

    void submitFeedback(const string &message) const {
        cout << "Feedback submitted by User ID: " << userId << " - " << message << endl;
    }
};

// Initialize static counter
int User::userCounter = 1;

class Service {
public:
    static void listAllServices() {
        cout << "1. Birth Certificate - Issuance of Birth Certificate" << endl;
        cout << "2. Driving License - Issuance of Driving License" << endl;
    }
};

class Application {
private:
    int applicationId;
    int userId;
    int serviceId;
    string applicationDate;
    string status;
    string documents;

public:
    Application(int appId, int userId, int serviceId, const string &appDate, const string &status, const string &documents)
        : applicationId(appId), userId(userId), serviceId(serviceId), applicationDate(appDate), status(status), documents(documents) {}

    void submitApplication() const {
        cout << "Application submitted successfully. Application ID: " << applicationId << endl;
    }

    void checkStatus() const {
        cout << "Application status: " << status << endl;
    }

    void uploadDocuments(const string &newDocuments) {
        documents = newDocuments;
        cout << "Documents uploaded: " << documents << endl;
    }

    void editApplication(const string &newStatus) {
        status = newStatus;
        cout << "Application status updated to: " << status << endl;
    }
};

class Payment {
private:
    int paymentId;
    int applicationId;
    double amount;
    string paymentDate;

public:
    Payment(int payId, int appId, double amount, const string &payDate)
        : paymentId(payId), applicationId(appId), amount(amount), paymentDate(payDate) {}

    void makePayment() const {
        cout << "Payment made successfully. Payment ID: " << paymentId << endl;
    }

    void getPaymentStatus() const {
        cout << "Payment status: Successful" << endl;
    }

    void generateReceipt() const {
        cout << "Receipt generated for Payment ID: " << paymentId << endl;
    }
};

class Notification {
private:
    int notificationId;
    int userId;
    string message;
    string date;

public:
    Notification(int notifId, int userId, const string &message, const string &date)
        : notificationId(notifId), userId(userId), message(message), date(date) {}

    void sendNotification(int userId, const string &message) const {
        cout << "Notification sent to User ID: " << userId << " - " << message << endl;
    }

    void markAsRead() const {
        cout << "Notification marked as read." << endl;
    }

    void getNotifications(int userId) const {
        cout << "Displaying notifications for User ID: " << userId << endl;
    }
};

class Feedback {
private:
    int feedbackId;
    int userId;
    string message;
    string date;

public:
    Feedback(int feedId, int userId, const string &message, const string &date)
        : feedbackId(feedId), userId(userId), message(message), date(date) {}

    void submitFeedback(int userId, const string &message) const {
        cout << "Feedback submitted by User ID: " << userId << " - " << message << endl;
    }

    void viewFeedback() const {
        cout << "Viewing feedback for Feedback ID: " << feedbackId << endl;
    }

    void responseToFeedback(const string &response) const {
        cout << "Response to feedback: " << response << endl;
    }
};

class SystemManager {
private:
    User* user;
    bool isLoggedIn;

    void showMenu() const {
        cout << "\n--- Welcome to Goa Online Services ---" << endl;
        cout << "1. Register User" << endl;
        cout << "2. Login" << endl;
        cout << "3. View Services" << endl;
        cout << "4. Submit Application" << endl;
        cout << "5. Make Payment" << endl;
        cout << "6. Send Notification" << endl;
        cout << "7. Submit Feedback" << endl;
        cout << "8. View Departments" << endl;
        cout << "9. Exit" << endl;
        cout << "Select an option: ";
    }

    void registerUser() {
        string name, email, phoneNumber, address, userType;
        cout << "Register a new user:" << endl;
        cout << "Enter Name: ";
        getline(cin, name);

        cout << "Enter Email: ";
        getline(cin, email);

        cout << "Enter Phone Number: ";
        getline(cin, phoneNumber);

        cout << "Enter Address: ";
        getline(cin, address);

        cout << "Enter User Type (Citizen/Admin): ";
        getline(cin, userType);

        user = new User(name, email, phoneNumber, address, userType);
        user->registerUser();
    }

    void loginUser() {
        if (user == nullptr) {
            cout << "No user registered. Please register first." << endl;
        } else {
            string email, phoneNumber;
            cout << "Login:" << endl;
            cout << "Enter Email: ";
            getline(cin, email);

            cout << "Enter Phone Number: ";
            getline(cin, phoneNumber);

            isLoggedIn = user->login(email, phoneNumber);
        }
    }

    void viewServices() const {
        if (isLoggedIn) {
            Service::listAllServices();
        } else {
            cout << "Please log in first." << endl;
        }
    }

    void submitApplication() const {
        if (isLoggedIn) {
            cout << "Submit an Application:" << endl;
            int applicationId, serviceId;
            string applicationDate, status, documents;

            cout << "Enter Application ID: ";
            cin >> applicationId;

            cout << "Enter Service ID: ";
            cin >> serviceId;
            cin.ignore(); // Ignore newline character

            cout << "Enter Application Date (YYYY-MM-DD): ";
            getline(cin, applicationDate);

            cout << "Enter Status (Pending/Approved): ";
            getline(cin, status);

            cout << "Enter Documents: ";
            getline(cin, documents);

            Application app(applicationId, user->getUserId(), serviceId, applicationDate, status, documents);
            app.submitApplication();
        } else {
            cout << "Please log in first." << endl;
        }
    }

    void makePayment() const {
        if (isLoggedIn) {
            cout << "Make a Payment:" << endl;
            int paymentId, applicationId;
            double amount;
            string paymentDate;

            cout << "Enter Payment ID: ";
            cin >> paymentId;

            cout << "Enter Application ID: ";
            cin >> applicationId;

            cout << "Enter Amount: ";
            cin >> amount;
            cin.ignore(); // Ignore newline character

            cout << "Enter Payment Date (YYYY-MM-DD): ";
            getline(cin, paymentDate);

            Payment payment(paymentId, applicationId, amount, paymentDate);
            payment.makePayment();
        } else {
            cout << "Please log in first." << endl;
        }
    }

    void sendNotification() const {
        if (isLoggedIn) {
            cout << "Send a Notification:" << endl;
            int notificationId;
            string message, date;

            cout << "Enter Notification ID: ";
            cin >> notificationId;
            cin.ignore(); // Ignore newline character

            cout << "Enter Message: ";
            getline(cin, message);

            cout << "Enter Date (YYYY-MM-DD): ";
            getline(cin, date);

            Notification notification(notificationId, user->getUserId(), message, date);
            notification.sendNotification(user->getUserId(), message);
        } else {
            cout << "Please log in first." << endl;
        }
    }

    void submitFeedback() const {
        if (isLoggedIn) {
            cout << "Submit Feedback:" << endl;
            int feedbackId;
            string message, date;

            cout << "Enter Feedback ID: ";
            cin >> feedbackId;
            cin.ignore(); // Ignore newline character

            cout << "Enter Message: ";
            getline(cin, message);

            cout << "Enter Date (YYYY-MM-DD): ";
            getline(cin, date);

            Feedback feedback(feedbackId, user->getUserId(), message, date);
            feedback.submitFeedback(user->getUserId(), message);
        } else {
            cout << "Please log in first." << endl;
        }
    }

public:
    SystemManager() : user(nullptr), isLoggedIn(false) {}

    void run() {
        int option;
        while (true) {
            showMenu();
            cin >> option;
            cin.ignore(); // Ignore newline character

            switch (option) {
                case 1:
                    registerUser();
                    break;
                case 2:
                    loginUser();
                    break;
                case 3:
                    viewServices();
                    break;
                case 4:
                    submitApplication();
                    break;
                case 5:
                    makePayment();
                    break;
                case 6:
                    sendNotification();
                    break;
                case 7:
                    submitFeedback();
                    break;
                case 8:
                    cout << "View Departments functionality not implemented yet." << endl;
                    break;
                case 9:
                    cout << "Exiting the system." << endl;
                    return;
                default:
                    cout << "Invalid option. Please try again." << endl;
            }
        }
    }
};

int main() {
    SystemManager manager;
    manager.run();
    return 0;
}

File: Python/Code/GoaOnline.py
class User:
    user_counter = 1

    def __init__(self, name, email, phone_number, address, user_type):
        self.user_id = User.user_counter
        User.user_counter += 1
        self.name = name
        self.email = email
        self.phone_number = phone_number
        self.address = address
        self.user_type = user_type

    def get_user_id(self):
        return self.user_id

    def register_user(self):
        print(f"User registered successfully: {self.name} (User ID: {self.user_id})")

    def login(self, email, phone_number):
        if self.email == email and self.phone_number == phone_number:
            print(f"User logged in successfully: {self.name}")
            return True
        else:
            print("Login failed. Incorrect email or phone number.")
            return False

    def update_profile(self, new_address):
        self.address = new_address
        print(f"Profile updated with new address: {self.address}")

    def view_services(self):
        print(f"Viewing available services for user: {self.name}")

    def submit_feedback(self, message):
        print(f"Feedback submitted by User ID: {self.user_id} - {message}")


class Service:
    @staticmethod
    def list_all_services():
        print("1. Birth Certificate - Issuance of Birth Certificate")
        print("2. Residence Certificate - Issuance of Residence Certificate")


class Application:
    def __init__(self, app_id, user_id, service_id, app_date, status, documents):
        self.application_id = app_id
        self.user_id = user_id
        self.service_id = service_id
        self.application_date = app_date
        self.status = status
        self.documents = documents

    def submit_application(self):
        print(f"Application submitted successfully. Application ID: {self.application_id}")

    def check_status(self):
        print(f"Application status: {self.status}")

    def upload_documents(self, new_documents):
        self.documents = new_documents
        print(f"Documents uploaded: {self.documents}")

    def edit_application(self, new_status):
        self.status = new_status
        print(f"Application status updated to: {self.status}")


class Payment:
    def __init__(self, pay_id, app_id, amount, pay_date):
        self.payment_id = pay_id
        self.application_id = app_id
        self.amount = amount
        self.payment_date = pay_date

    def make_payment(self):
        print(f"Payment made successfully. Payment ID: {self.payment_id}")

    def get_payment_status(self):
        print("Payment status: Successful")

    def generate_receipt(self):
        print(f"Receipt generated for Payment ID: {self.payment_id}")


class Notification:
    def __init__(self, notif_id, user_id, message, date):
        self.notification_id = notif_id
        self.user_id = user_id
        self.message = message
        self.date = date

    def send_notification(self, user_id, message):
        print(f"Notification sent to User ID: {user_id} - {message}")

    def mark_as_read(self):
        print("Notification marked as read.")

    def get_notifications(self, user_id):
        print(f"Displaying notifications for User ID: {user_id}")


class Feedback:
    def __init__(self, feed_id, user_id, message, date):
        self.feedback_id = feed_id
        self.user_id = user_id
        self.message = message
        self.date = date

    def submit_feedback(self, user_id, message):
        print(f"Feedback submitted by User ID: {user_id} - {message}")

    def view_feedback(self):
        print(f"Viewing feedback for Feedback ID: {self.feedback_id}")

    def response_to_feedback(self, response):
        print(f"Response to feedback: {response}")


class SystemManager:
    def __init__(self):
        self.user = None
        self.is_logged_in = False

    def show_menu(self):
        print("\n--- Welcome to Goa Online Services ---")
        print("1. Register User")
        print("2. Login")
        print("3. View Services")
        print("4. Submit Application")
        print("5. Make Payment")
        print("6. Send Notification")
        print("7. Submit Feedback")
        print("8. View Departments")
        print("9. Exit")
        option = int(input("Select an option: "))
        return option

    def register_user(self):
        print("Register a new user:")
        name = input("Enter Name: ")
        email = input("Enter Email: ")
        phone_number = input("Enter Phone Number: ")
        address = input("Enter Address: ")
        user_type = input("Enter User Type (Citizen/Admin): ")
        self.user = User(name, email, phone_number, address, user_type)
        self.user.register_user()

    def login_user(self):
        if self.user is None:
            print("No user registered. Please register first.")
        else:
            email = input("Enter Email: ")
            phone_number = input("Enter Phone Number: ")
            self.is_logged_in = self.user.login(email, phone_number)

    def view_services(self):
        if self.is_logged_in:
            Service.list_all_services()
        else:
            print("Please log in first.")

    def submit_application(self):
        if self.is_logged_in:
            print("Submit an Application:")
            app_id = int(input("Enter Application ID: "))
            service_id = int(input("Enter Service ID: "))
            app_date = input("Enter Application Date (YYYY-MM-DD): ")
            status = input("Enter Status (Pending/Approved): ")
            documents = input("Enter Documents: ")
            app = Application(app_id, self.user.get_user_id(), service_id, app_date, status, documents)
            app.submit_application()
        else:
            print("Please log in first.")

    def make_payment(self):
        if self.is_logged_in:
            print("Make a Payment:")
            pay_id = int(input("Enter Payment ID: "))
            app_id = int(input("Enter Application ID: "))
            amount = float(input("Enter Amount: "))
            pay_date = input("Enter Payment Date (YYYY-MM-DD): ")
            payment = Payment(pay_id, app_id, amount, pay_date)
            payment.make_payment()
        else:
            print("Please log in first.")

    def send_notification(self):
        if self.is_logged_in:
            print("Send a Notification:")
            notif_id = int(input("Enter Notification ID: "))
            message = input("Enter Message: ")
            date = input("Enter Date (YYYY-MM-DD): ")
            notification = Notification(notif_id, self.user.get_user_id(), message, date)
            notification.send_notification(self.user.get_user_id(), message)
        else:
            print("Please log in first.")

    def submit_feedback(self):
        if self.is_logged_in:
            print("Submit Feedback:")
            feed_id = int(input("Enter Feedback ID: "))
            message = input("Enter Message: ")
            date = input("Enter Date (YYYY-MM-DD): ")
            feedback = Feedback(feed_id, self.user.get_user_id(), message, date)
            feedback.submit_feedback(self.user.get_user_id(), message)
        else:
            print("Please log in first.")

    def run(self):
        while True:
            option = self.show_menu()
            if option == 1:
                self.register_user()
            elif option == 2:
                self.login_user()
            elif option == 3:
                self.view_services()
            elif option == 4:
                self.submit_application()
            elif option == 5:
                self.make_payment()
            elif option == 6:
                self.send_notification()
            elif option == 7:
                self.submit_feedback()
            elif option == 8:
                print("View Departments functionality not implemented yet.")
            elif option == 9:
                print("Exiting the system.")
                break
            else:
                print("Invalid option. Please try again.")


if __name__ == "__main__":
    manager = SystemManager()
    manager.run()
